### PR TITLE
Brief design for summarizing information necessary for code generators in SQLFlow

### PR DESCRIPTION
**Pull request by uuleon**

### CODE REVIEW

File: doc/train_parameters.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# Brief design for unifying format of training parameters in SQLFlow
+
+SQLFlow propose an extension to the SQL SELECT statement, which proposes the syntax for training as
```
Comment (wangkuiyi) [inline doc/train_parameters.md]: SQLFlow extends the syntax of the SELECT statement of SQL to support training a model:

File: doc/train_parameters.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+TRAIN DNNClassifier       /* a pre-defined TensorFlow estimator, tf.estimator.DNNClassifier */
+WITH layers=[100, 200],   /* a parameter of the Estimator class constructor */
+     train.batch_size = 8 /* a parameter of the Estimator.train method */
+COLUMN *,                 /* all columns as raw features */
+       cross(v1, v9, v28) /* plus a derived (crossed) column */
+LABEL class
+INTO sqlflow_models.my_model_table;      /* saves trained model parameters and features into a table */
+```
+
+The syntax for training must following rules as
```
Comment (wangkuiyi) [inline doc/train_parameters.md]: Currently, we have the following syntax allowing users to provide necessary information for the training.

File: doc/train_parameters.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+```sql
+SELECT STATEMENT
+TRAIN ...
+WITH ...
+COLUMN ... [FOR ...]
+LABEL ...
+INTO ...
+```
+
+The training parameters in `WITH` or `COLUMN` is arbitrary for the `parser` of SQLFlow and validated in `codegen.go` or `codegen_alps.go` which may be using different and even conflicting rules.
```
Comment (wangkuiyi) [inline doc/train_parameters.md]: SQLFlow server passes the above information to code generators like `sql/codegen.go` and `sql/codegen_alps.go`, which generates the training program.

File: doc/train_parameters.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+Things will be even more difficult if we got other kinds of `codegen_**.go` in the future. 
+
+This document aims to provide a set of unifying format for training parameters.
```
Comment (wangkuiyi) [inline doc/train_parameters.md]: In this document, we summarize information necessary for the code generators.

File: doc/train_parameters.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+## Unifying format of training parameters
```
Comment (wangkuiyi) [inline doc/train_parameters.md]:
```
## Necessary Information for Training
```

File: doc/train_parameters.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+```
+represents that the `c1` field is the dense format and the `c2` field is the sparse format.
+
+<b>Here is a list of supported expression:</b>
+
+| expression | arguments of expression                                          | example                            |
+|------------|------------------------------------------------------------------|------------------------------------|
+| dense      | 1. field name (str) <br> 2. dense shape (integer) <br> 3. separator (str)  | dense(c1, 100, comma)              |
+| sparse     | 1. field name (str) <br> 2. sparse shape (integer) <br> 3. separator (str) | sparse(c2, 10000, comma)           |
+| numeric    | 1. field name (str) <br> 2. shape (integer)                           | numeric(c1, 100)                   |
```
Comment (typhoonzero) [inline doc/train_parameters.md:116]: What's the shape of `numeric` column, should it always be a scalar?  Actual training shape should be `(batch_size, 1)`

File: doc/train_parameters.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+| bucket     | 1. key (numeric) <br> 2. bucket size (integer)                        | bucket(numeric(c1, 100), 20)       |
+| cat_id     | 1. field name (str) <br> 2. bucket size (integer)                     | cat_id(c2, 10000)                  |
```
Comment (typhoonzero) [inline doc/train_parameters.md:118]: cat_id -> categorize

and, why always need bucket size, the column may be enumerated, like `Male, Female, NULL`

File: doc/train_parameters.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+| embedding  | 1. key (cat_id) <br> 2. dimension (integer) <br> 3. combiner (str)         | embedding(cat_id(c2, 10000), mean) |
+
```
Comment (typhoonzero) [inline doc/train_parameters.md:120]: should we also support `sequence(filed, padding_size)`?

File: doc/train_parameters.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	SPARSE(c2, ...)
+```
+represents that the `c1` field is the dense format and the `c2` field is the sparse format.
+
+<b>Here is a list of supported expression:</b>
+
+| expression | arguments of expression                                          | example                            |
+|------------|------------------------------------------------------------------|------------------------------------|
+| dense      | 1. field name (str) <br> 2. dense shape (integer) <br> 3. separator (str)  | dense(c1, 100, comma)              |
+| sparse     | 1. field name (str) <br> 2. sparse shape (integer) <br> 3. separator (str) | sparse(c2, 10000, comma)           |
```
Comment (typhoonzero) [inline doc/train_parameters.md:115]: maybe add data samples to indicate the column type?

File: doc/train_parameters.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+2. Fast fail during resolving before the code generation.
+
+Here we propose a module such as `resolving.go` which take the `trainClause` as input and outputs a struct named `resolvedTrainClause` for all kinds of `codegen_**.go` to do code generation.
+
+The struct named `resolvedTrainClause` looks like this
+```go
+package sql
+
+type resolvedTrainClause struct {
+	IsPreMadeModel  bool
```
Comment (typhoonzero) [inline doc/train_parameters.md:136]: We would also like to support keras models, and the generated code could have a lot of difference than estimator code. Will the "resolver" also consider dealing with keras models?

File: doc/train_parameters.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+
+## Independent module for resolving of training parameters
+Since the format of training parameters has been unified, it's better to have an independent module in SQLFlow to do the resolving according to the rules instead of doing it in each `codegen_**.go`.
+
+The advantage of an independent module contains
+1. Unifying resolving process to avoid differences in processing between `codegen` modules. 
+2. Fast fail during resolving before the code generation.
+
+Here we propose a module such as `resolving.go` which take the `trainClause` as input and outputs a struct named `resolvedTrainClause` for all kinds of `codegen_**.go` to do code generation.
```
Comment (typhoonzero) [inline doc/train_parameters.md:129]: `resolving.go` is not telling what to resolve, maybe `resolveTrainClause.go` is better?

File: doc/train_parameters.md (hunk 7 above)
Comment (uuleon) [reply doc/train_parameters.md:118]: In the original design, the `cat_id` is short for `categorical_column_with_identity ` which needs `bucket size`. Does renaming to `categorize` means it has to support all the APIs of `categorical_column_*`?

File: doc/train_parameters.md (hunk 6 above)
Comment (uuleon) [reply doc/train_parameters.md:116]: It should be a list of integer, thanks.

File: doc/train_parameters.md (hunk 11 above)
Comment (uuleon) [reply doc/train_parameters.md:129]: I was thinking maybe to support resolving `predictClause` in the future.

File: doc/train_parameters.md (hunk 9 above)
Comment (uuleon) [reply doc/train_parameters.md:115]: It is in the `sqlflow/doc/alps_submitter.md`.

File: doc/train_parameters.md (hunk 1 above)
Comment (uuleon) [reply doc/train_parameters.md]: Thanks, fixed.

**Review by typhoonzero — APPROVED** [review]
I am approving this so that later work can move on.

**Pull request merged**